File: src/main/java/com/github/simiacryptus/aicoder/actions/generic/GenericAppend.java
package com.github.simiacryptus.aicoder.actions.generic;

import com.github.simiacryptus.aicoder.config.AppSettingsState;
import com.github.simiacryptus.aicoder.openai.CompletionRequest;
import com.github.simiacryptus.aicoder.util.ComputerLanguage;
import com.github.simiacryptus.aicoder.util.UITools;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Document;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;


/**
 * The GenericAppend IntelliJ action allows users to quickly append a prompt to the end of a selected text.
 * To use, select some text and then select the GenericAppend action from the editor context menu.
 * The action will insert the completion at the end of the selected text.
 */
public class GenericAppend extends AnAction {

    @Override
    public void update(@NotNull AnActionEvent e) {
        e.getPresentation().setEnabledAndVisible(isEnabled(e));
        super.update(e);
    }

    @SuppressWarnings("unused")
    private static boolean isEnabled(@NotNull AnActionEvent e) {
        if (!UITools.hasSelection(e)) return false;
        return true;
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent event) {
        @Nullable Caret caret = event.getData(CommonDataKeys.CARET);
        @Nullable CharSequence before = Objects.requireNonNull(caret).getSelectedText();
        AppSettingsState settings = AppSettingsState.getInstance();
        @NotNull CompletionRequest completionRequest = settings.createCompletionRequest().appendPrompt(before);
        @NotNull Document document = event.getRequiredData(CommonDataKeys.EDITOR).getDocument();
        int selectionEnd = caret.getSelectionEnd();
        UITools.redoableRequest(completionRequest, "", event,
                newText -> UITools.insertString(document, selectionEnd, newText));
    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/actions/generic/GenericAppend.kt
package com.github.simiacryptus.aicoder.actions.generic

import com.github.simiacryptus.aicoder.config.AppSettingsState
import com.github.simiacryptus.aicoder.util.UITools.hasSelection
import com.github.simiacryptus.aicoder.util.UITools.insertString
import com.github.simiacryptus.aicoder.util.UITools.redoableRequest
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import java.util.*

/**
 * The GenericAppend IntelliJ action allows users to quickly append a prompt to the end of a selected text.
 * To use, select some text and then select the GenericAppend action from the editor context menu.
 * The action will insert the completion at the end of the selected text.
 */
class GenericAppend : AnAction() {
    override fun update(e: AnActionEvent) {
        e.presentation.isEnabledAndVisible = isEnabled(e)
        super.update(e)
    }

    override fun actionPerformed(event: AnActionEvent) {
        val caret = event.getData(CommonDataKeys.CARET)
        val before: CharSequence? = Objects.requireNonNull(caret)!!.selectedText
        val settings = AppSettingsState.getInstance()
        val completionRequest = settings.createCompletionRequest().appendPrompt(before)
        val document = event.getRequiredData(CommonDataKeys.EDITOR).document
        val selectionEnd = caret!!.selectionEnd
        redoableRequest(
            completionRequest, "", event
        ) { newText: CharSequence? ->
            insertString(
                document, selectionEnd,
                newText!!
            )
        }
    }

    companion object {
        @Suppress("unused")
        private fun isEnabled(e: AnActionEvent): Boolean {
            return hasSelection(e)
        }
    }
}





